accepted/cfi-improvements-with-pauth-and-bti.md

```diff
@@ -0,0 +1,142 @@
+start with the `PACIASP` instruction. However, the regular return will be preserved and instead will
+be preceded by the `AUTIASP` instruction. The reason is that both `AUTIASP` and `PACIASP` act as
+`NOP` instructions when executed by a processor that does not support PAuth, thus making the
+assembly code generic.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Since the existing implementation already uses the standard back-edge CFI techniques that are
+preferred in the absence of special hardware support (i.e. a separate protected stack that is not
+used for buffers that could be accessed out of bounds), the alternative is not to implement the
+proposal, so the rationale is based mainly on the overhead being insignificant. In terms of code
+size the impact of the back-edge CFI improvements is an additional instruction per function, or 2
+for functions implemented in assembly.
+
+The [Clang CFI design][clang-cfi-design] provides an idea for an alternative implementation of the
+forward-edge CFI mechanism that is enabled by BTI. It involves instrumenting every indirect branch
+to check if its destination is permitted. While the overhead of this approach can be reduced by
+using efficient data structures for the destination address lookup and optionally limiting the
+checks only to indirect function calls, it is still significantly larger than the worst-case BTI
+overhead of one instruction per basic block per function. On the other hand, it does not require any
+special hardware support, so it could be applied to all supported platforms.
+
+[clang-cfi-design]: https://clang.llvm.org/docs/ControlFlowIntegrityDesign.html
+
+# Open questions
+[open-questions]: #open-questions
+
+- What is the performance overhead of the proposal?
+- What technologies are available in other instruction set architectures to achieve the same goals?
+- What hardening approaches are applicable to the fiber implementation? The fiber switching code
+saves the values of all callee-saved registers on the stack, i.e. memory that is potentially
+accessible to an attacker. Some of those values could be code addresses that would be used by
+indirect branches, so should we devise a scheme to authenticate them? While the regular pointer
+authentication instructions assume that they are operating on valid virtual addresses (which implies
+that the most significant bits are redundant and could be repurposed), PAuth provides operations to
+authenticate arbitrary data, which could be used in this case.
+- Should we generate the operations that act as `NOP` instructions unconditionally instead (while
+still choosing the shorter alternative sequences if the target supports them)? That would
+especially help the ahead of time compilation use case, and could arguably reduce the amount of
+testing, i.e. no need to check both with and without CFI enhancements.
```

Comment:
I have changed the text, so that now the usage of both PAuth and BTI is disabled by default, and the values of the respective configuration options are decoupled from the capabilities of the host CPU.